### PR TITLE
python312Packages.quantities: prevent arbitrary code eval

### PR DESCRIPTION
**Pull request by dotlambda**
## Description of changes
applies https://github.com/python-quantities/python-quantities/pull/236

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by doronbehar — APPROVED** [review]
Diff LGTM.

**Comment by dotlambda** [comment]
@doronbehar Btw I pinged you because

> Added support for NumPy 2.0, while maintaining support for older versions back to 1.22

**Comment by dotlambda** [comment]
@ofborg eval

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 339916` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>python311Packages.neo</li>
    <li>python311Packages.neo.dist</li>
    <li>python312Packages.neo</li>
    <li>python312Packages.neo.dist</li>
  </ul>
</details>
<details>
  <summary>8 packages built:</summary>
  <ul>
    <li>python311Packages.pylatex</li>
    <li>python311Packages.pylatex.dist</li>
    <li>python311Packages.quantities</li>
    <li>python311Packages.quantities.dist</li>
    <li>python312Packages.pylatex</li>
    <li>python312Packages.pylatex.dist</li>
    <li>python312Packages.quantities</li>
    <li>python312Packages.quantities.dist</li>
  </ul>
</details>

**Comment by dotlambda** [comment]
https://github.com/NeuralEnsemble/python-neo/issues/1534

**Comment by risicle** [comment]
Think we might be better patching 24.05 instead of bumping, there could be a few compatibility risks with the other changes.

**Comment by risicle** [comment]
@ofborg build python312Packages.neo

looks like it needs its deps relaxed

**Comment by doronbehar** [comment]
> @ofborg build python312Packages.neo
> 
> looks like it needs its deps relaxed

But many tests fail when that is relaxed.

**Comment by doronbehar** [comment]
hmm merge conflict...

**Comment by doronbehar** [comment]
I think that because this is a security issue being fixed with the update we should merge this without waiting for neo to accommodate for quantities 0.16.0.

**Comment by dotlambda** [comment]
Result of `nixpkgs-review pr 339916` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>12 packages built:</summary>
  <ul>
    <li>python311Packages.neo</li>
    <li>python311Packages.neo.dist</li>
    <li>python311Packages.pylatex</li>
    <li>python311Packages.pylatex.dist</li>
    <li>python311Packages.quantities</li>
    <li>python311Packages.quantities.dist</li>
    <li>python312Packages.neo</li>
    <li>python312Packages.neo.dist</li>
    <li>python312Packages.pylatex</li>
    <li>python312Packages.pylatex.dist</li>
    <li>python312Packages.quantities</li>
    <li>python312Packages.quantities.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/quantities/default.nix [hunk 1]
```diff
@@ -12,21 +13,31 @@
...
+    repo = "python-quantities";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-N20xfzGtM0VnfkJtzMytNLySTkgVz2xf1nEJxlwBSCI=";
   };
 
-  nativeBuildInputs = [
+  patches = [
+    (fetchpatch2 {
+      name = "prevent-arbitrary-code-eval.patch";
+      url = "https://github.com/python-quantities/python-quantities/pull/236.patch";
```
Comment (risicle) [inline pkgs/development/python-modules/quantities/default.nix:30]:
```suggestion
      url = "https://github.com/python-quantities/python-quantities/commit/329433cac3776d58b208952e328014f716cdfa92.patch";
```

Direct commit references are usually preferred (hash unchanged in this case)

Comment (dotlambda) [reply pkgs/development/python-modules/quantities/default.nix:30]: Since the PR has been merged this is fine.

**Comment by doronbehar** [comment]
@bcdarwin please note that the update to quantities 0.16.0 will break `python312Packages.neo` which you maintain, see upstream issue:

https://github.com/NeuralEnsemble/python-neo/issues/1534

**Pull request merged**